config/config/src/main/java/io/helidon/config/EnumMapperProvider.java

```diff
@@ -0,0 +1,113 @@
+ *         <li>If the matcher finds a <em>case-sensitive</em> match with an enum value name, then that enum value matches.</li>
+ *         <li>If the matcher finds exactly one <em>case-insensitive</em> match, that enum value matches.</li>
+ *         <li>If the matcher finds no matches or multiple matches, throw a
+ *         {@link io.helidon.config.ConfigEnumMappingException} with a message explaining the problem.</li>
+ *     </ul>
+ *     These conversions are intended to maximize ease-of-use for authors of config sources so the values need not be
+ *     upper-cased nor punctuated with underscores rather than the more conventional (in config at least) hyphen.
+ * </p>
+ * <p>
+ *     The only hardship this imposes is if a confusingly-designed enum has values which differ only in case <em>and</em> the
+ *     string in the config source does not exactly match one of the enum value names. In such cases
+ *     the mapper will be unable to choose which enum value matches an ambiguous string. A developer faced with this
+ *     problem can simply provide her own explicit config mapping for that enum, for instance as a function parameter to
+ *     {@code Config#as}.
+ * </p>
+ *
+ */
+class EnumMapperProvider implements ConfigMapperProvider {
+
+    private final Map<Class<?>, Function<Config, ?>> mappers = new HashMap<>();
+
+    @Override
+    public Map<Class<?>, Function<Config, ?>> mappers() {
+        return mappers;
+    }
+
+    @Override
+    public <T> Optional<Function<Config, T>> mapper(Class<T> type) {
+        if (!type.isEnum()) {
+            return Optional.empty();
+        }
+
+        Function<Config, T> result = (Function<Config, T>) mappers.computeIfAbsent(type, this::enumMapper);
+        return Optional.of(result);
+    }
+
+    private <T> Function<Config, T> enumMapper(Class<T> type) {
+        return config -> {
+            // type.isEnum() must be true in the caller, so we know the type is Class<Enum<?>>.
+            Class<Enum<?>> enumType = (Class<Enum<?>>) type;
+            if (!config.hasValue() || !config.exists()) {
+                throw MissingValueException.create(config.key());
+            }
+            if (!config.isLeaf()) {
+                throw new ConfigMappingException(config.key(), enumType, "config node must be a leaf but is not");
+            }
+            String value = config.asString().get();
+            String convertedValue = value.replace('-', '_');
+            List<Enum<?>> caseInsensitiveMatches = new ArrayList<>();
+            for (Enum<?> candidate : enumType.getEnumConstants()) {
+                // Check for an exact match first, with or without hyphen conversion.
+                if (candidate.name().equals(convertedValue) || candidate.name().equals(value)) {
+                    return (T) candidate;
+                }
+                if (candidate.name().compareToIgnoreCase(convertedValue) == 0
```

Comment:
why do you use `compareToIgnoreCase` when there is `equalsIgnoreCase`?

Reply:
No particular reason. Will change.